### PR TITLE
Dynect Provider does not returns type-safe rdata

### LINKED ISSUE #296: Dynect Provider does not returns type-safe rdata
in Dynect Provider, the ResourceRecordSet returned by api, always contains records of type LinkedHashMap. It should contain records of type-safe subclasses like AData, AAAAData etc. All other providers return type-safe rdata.

### PR DESCRIPTION
**Pull request by santhosh-tekuri**
in Dynect Provider, the ResourceRecordSet returned by api, always
contains records of type LinkedHashMap. It should contain records of
type-safe subclasses like AData, AAAAData etc. All other providers
return type-safe rdata.

closes #296

### CODE REVIEW

**Comment by adriancole** [comment]
Looks good, but I would guess you would want a test case, right? I don't recall if dynect has an existing test class for rdata conversions, but I know some providers do.

Ps when you are raising both the issue and the solution, just open a PR (no tracking issue needed).

**Comment by santhosh-tekuri** [comment]
I did not find any test cases for rdata conversions in other providers. could you point to any, if you know.

in case if i open just PR: then i dont need to add "closes #xxx" in comment. am i right?

**Comment by adriancole** [comment]
>
> I did not find any test cases for rdata conversions in other providers.
> could you point to any, if you know.
>
Here's an example:

https://github.com/Netflix/denominator/blob/master/ultradns/src/test/java/denominator/ultradns/UltraDNSFunctionsTest.java


> in case if i open just PR: then i dont need to add "closes #xxx" in
> comment. am i right?
>
Yep!

File: dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package denominator.dynect;
+
+import com.google.gson.*;
```
Comment (adriancole) [inline dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java]: nit: no wildcards, please!

**Comment by santhosh-tekuri** [comment]
i am using Idea, and it puts them;

also i manually changed from 4space indentation to 2space etc....

File: dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Map;
+
+import static org.testng.Assert.assertEquals;
+
+public class DynECTFunctionsTest{
```
Comment (adriancole) [inline dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java]: I understand that you are trying to not repeat yourself, but this class is very indirect. Unit tests should favor readability and straight-forwardness above line count. When we get meta, we lose the purpose of the file and force users to scroll through multiple files and boilerplate concepts to understand what's going on. This goes against the direction of the work I've been on, which has been to simplify how our tests behave.

Try the following, which is very straight-forward and requires no scrolling.

```java
  @Test
  public void aaaaRecord() throws Exception {
    assertThat(toRData("{\n"
        + "    \"zone\": \"denominator.io\",\n"
        + "    \"ttl\": 3600,\n"
        + "    \"fqdn\": \"www.denominator.io\",\n"
        + "    \"record_type\": \"AAAA\",\n"
        + "    \"rdata\": {\n"
        + "      \"address\": \"2001:db8::3\"\n"
        + "    },\n"
        + "    \"record_id\": 2\n"
        + "  }"))
        .isInstanceOf(AAAAData.class)
        .containsEntry("address", "2001:db8::3")
    // or
        .isEqualTo(AAAAData.create("2001:db8::3"));
  }

  private static Map<String, Object> toRData(String inputJson) {
    return ToRecord.toRData(new JsonParser().parse(inputJson).getAsJsonObject());
  }
```

Or the more meta, but still much less difficult to understand parameterized version

```java

  @DataProvider(name = "rdata") public static Object[][] rdata() {
    return new Object[][] {
        {
            "{\n"
          + "    \"zone\": \"denominator.io\",\n"
          + "    \"ttl\": 3600,\n"
          + "    \"fqdn\": \"www.denominator.io\",\n"
          + "    \"record_type\": \"AAAA\",\n"
          + "    \"rdata\": {\n"
          + "      \"address\": \"2001:db8::3\"\n"
          + "    },\n"
          + "    \"record_id\": 2\n"
          + "  }",
            AAAAData.create("2001:db8::3")
        } // repeat for each rdata to test
    };
  }

  @Test(dataProvider = "rdata")
  public void toRData(String inputJson, Map<String, Object> expectedResult) {
    assertThat(ToRecord.toRData(new JsonParser().parse(inputJson).getAsJsonObject()))
        .isInstanceOf(expectedResult.getClass())
        .isEqualTo(expectedResult);
  }
```

Hope this helps!

**Comment by adriancole** [comment]
If you use this, it should correct the import rules.

http://code.google.com/p/google-styleguide/source/browse/trunk/intellij-java-google-style.xml

File: dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import denominator.model.rdata.SPFData;
+import denominator.model.rdata.SRVData;
+import denominator.model.rdata.TXTData;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class DynECTFunctionsTest{
+
+  @DataProvider(name = "rdata")
+  public Object[][] rdata() throws IOException {
```
Comment (adriancole) [inline dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java]: pretty!

**Comment by adriancole** [comment]
looks great, Santosh! just reword the commit and add to changelog:

`Changes DynECT to return type-safe rdata`

**Pull request merged**